### PR TITLE
cmd/oci-image-tool: add manuals

### PR DESCRIPTION
**Pull request by runcom**
Signed-off-by: Antonio Murdaca <runcom@redhat.com>

### CODE REVIEW

**Comment by wking** [comment]
If we're getting more formal about documenting oci-image-tool, I
recommend following ocitools and putting most of the docs in something
that can turn into a man page [1].

[1]: https://github.com/opencontainers/ocitools/tree/master/man

**Comment by runcom** [comment]
> If we're getting more formal about documenting oci-image-tool, I
> recommend following ocitools and putting most of the docs in something
> that can turn into a man page [1].

if maintainers are with that, I'll do that - right now I don't think it's necessary, `ocitools` otoh is already used a lot and more mature.

**Comment by wking** [comment]
On Fri, Jul 22, 2016 at 07:49:41AM -0700, Antonio Murdaca wrote:
> > If we're getting more formal about documenting oci-image-tool, I
> > recommend following ocitools and putting most of the docs in something
> > that can turn into a man page [1].
>
> if maintainers are with that, I'll do that - right now I don't think
> it's necessary, `ocitools` otoh is already used a lot and more
> mature.

Hopefully oci-image-tool will be mature and well-used soon ;).
Starting with a man page saves us from having to identify that point
and transition the README to a man page later.

**Comment by runcom** [comment]
> and transition the README to a man page later.

I'd still have a README - why not?

**Comment by wking** [comment]
On Fri, Jul 22, 2016 at 07:58:11AM -0700, Antonio Murdaca wrote:
> > and transition the README to a man page later.
> 
> I'd still have a README - why not?

A README should defer to the man-page (opencontainers/ocitools#47) to
stay DRY.

**Comment by runcom** [comment]
> A README should defer to the man-page (opencontainers/ocitools#47) to
> stay DRY.

but we don't have a man page yet and what I've added is dry enough for now - I can remove the Usage to be referenced in future man pages but examples are good there and ocitools have them as well

**Comment by wking** [comment]
On Fri, Jul 22, 2016 at 08:07:10AM -0700, Antonio Murdaca wrote:
> > A README should defer to the man-page (opencontainers/ocitools#47) to
> > stay DRY.
>
> but we don't have a man page yet and what I've added is dry enough
> for now…

I'm just saying that if the decision is “we don't want a man page yet
because $REASONS”, we want to be clear on those reasons so we know
when to re-evaluate.  If the issue is “because I don't want to put in
the time to set that up”, I'm happy to file a man-page conversion
against this branch later today.

**Comment by runcom** [comment]
> If the issue is “because I don't want to put in
> the time to set that up”, I'm happy to file a man-page conversion
> against this branch later today.

that's not the issue - I'm only saying a README.md with examples can go in w/o waiting to have man pages and that can come as a follow on (by me as well of course)

**Comment by wking** [comment]
What examples would you put in a README that wouldn't go in the man page?

**Comment by runcom** [comment]
> What examples would you put in a README that wouldn't go in the man page?

I'm not saying they should not go in the man page - I'm saying this is adding some example the same as the https://github.com/opencontainers/ocitools README is doing (https://github.com/opencontainers/ocitools/blob/master/man/ocitools-validate.1.md doesn't have examples in it but the README does, so take this PR as the README side of what ocitools already contain)

as said, I'll add some man pages stub

**Comment by wking** [comment]
On Fri, Jul 22, 2016 at 08:31:32AM -0700, Antonio Murdaca wrote:
> > What examples would you put in a README that wouldn't go in the
> > man page?
>
> I'm not saying they should not go in the man page - I'm saying this
> is adding some example the same as the
> https://github.com/opencontainers/ocitools README is doing
> (https://github.com/opencontainers/ocitools/blob/master/man/ocitools-validate.1.md
> doesn't have examples in it but the README does…

Ah, I think that should be fixed in ocitools.  Filed as
opencontainers/ocitools#142.

**Comment by runcom** [comment]
man pages added...

File: cmd/oci-image-tool/man/oci-image-tool-unpack.1.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+% OCI Community
+% JULY 2016
+# NAME
+oci-image-tool-unpack \- Unpack an image or image source layout
+
+# SYNOPSIS
+**oci-image-tool unpack** [src] [dest] [flags]
+
+# DESCRIPTION
+`oci-image-tool unpack` unpacks a given OCI image into a directory suitable to be used with `runc`.
```
Comment (wking) [inline cmd/oci-image-tool/man/oci-image-tool-unpack.1.md]: It's probably worth drawing a clear distinction between `unpack` and `create-runtime-bundle` here.  That distinction is mostly around rootfs-or-not and config translation.

Comment (runcom) [inline cmd/oci-image-tool/man/oci-image-tool-unpack.1.md]: sure, any suggestion on the language?

**Comment by wking** [comment]
On Fri, Jul 22, 2016 at 10:13:40AM -0700, Antonio Murdaca wrote:
> man pages added...

They look pretty good to me.  Do we expect additional command-line
tools in image-spec?  If so, Makefiles for building the man pages will
be simpler if we use man/ instead of cmd/oci-image-tool/man/.

**Comment by runcom** [comment]
> If so, Makefiles for building the man pages will
> be simpler if we use man/ instead of cmd/oci-image-tool/man/.

I put them there just in case we'll have a dedicated repo for oci-image-tool and generally to keep them there not to clutter the repo root

File: cmd/oci-image-tool/man/oci-image-tool-unpack.1.md (hunk 1 above)
Comment (wking) [inline cmd/oci-image-tool/man/oci-image-tool-unpack.1.md]: On Fri, Jul 22, 2016 at 10:25:35AM -0700, Antonio Murdaca wrote:
> > +# DESCRIPTION
> > +`oci-image-tool unpack` unpacks a given OCI image into a directory suitable to be used with `runc`.
> 
> sure, any suggestion on the language?

For unpack, something like:

  Validates an application/vnd.oci.image.manifest.v1+json and unpacks
  its layered filesystem to DEST.

For create-runtime-bundle, something like:

  Validates an application/vnd.oci.image.manifest.v1+json and unpacks
  its layered filesystem to DEST/rootfs, although the target directory
  is configurable with --rootfs.  See oci-image-tool-unpack(1) for
  more details on this process.

  Also translates the referenced config from
  application/vnd.oci.image.serialization.config.v1+json to a
  runtime-spec's config.json.

**Comment by wking** [comment]
On Fri, Jul 22, 2016 at 10:31:26AM -0700, Antonio Murdaca wrote:
> > If so, Makefiles for building the man pages will
> > be simpler if we use man/ instead of cmd/oci-image-tool/man/.
> 
> I put them there just in case we'll have a dedicated repo for
> oci-image-tool…

+1 to that [1] ;).

[1]: https://github.com/opencontainers/ocitools/issues/83#issuecomment-222886302

File: cmd/oci-image-tool/man/oci-image-tool-unpack.1.md (hunk 1 above)
Comment (wking) [inline cmd/oci-image-tool/man/oci-image-tool-unpack.1.md]: On Fri, Jul 22, 2016 at 10:35:50AM -0700, W. Trevor King wrote:
>   Also translates the referenced config from
>   application/vnd.oci.image.serialization.config.v1+json to a
>   runtime-spec's config.json.

Probably worth working in DEST/config.json there somewhow.  Maybe:

  Also translates the referenced config from
  application/vnd.oci.image.serialization.config.v1+json to a
  runtime-spec-compatible DEST/config.json.

**Comment by vbatts** [comment]
LGTM

**Comment by philips** [comment]
Is this generated somehow? Ideally the help text and the docs stay in sync.

In any case:
LGTM

**Comment by wking** [comment]
@runcom, would you like to file a follow-up with the details on unpack
vs. create-runtime-bundle [1] or shall I?

[1]: https://github.com/opencontainers/image-spec/pull/180/files#r71914328

**Comment by runcom** [comment]
I will @wking

**Comment by wking** [comment]
On Wed, Jul 27, 2016 at 05:30:08PM -0700, Antonio Murdaca wrote:
> I will @wking

Any progress on this?  I'm happy to turn [1] into a PR if you're too
busy.

[1]: https://github.com/opencontainers/image-spec/pull/180/files#r71914328

**Comment by philips** [comment]
@runcom @wking can you file a new issue so we don't lose this?

**Comment by wking** [comment]
On Wed, Aug 31, 2016 at 10:53:18AM -0700, Brandon Philips wrote:
> @runcom @wking can you file a new issue so we don't lose this?

I'll give @runcom 24 hours to get back, otherwise I'll just file a PR
;).

**Pull request merged**